### PR TITLE
Add badge model and database interaction

### PR DESCRIPTION
**Pull request by Netzu00**
So far, the badge model has been added, and a test badge can be added to the user's badge collection in the database by clicking the ScribeWiz logo button in profile page. The intended use of the button is to bring up a window that displays the user's badge collection, which will be added eventually.

### CODE REVIEW

**Comment by Netzu00** [comment]
Added the badge collection window, which displays the badges associated to the currently logged in user (badge data taken from the database). Badges have 3 different ranks (Gold, Silver, Bronze), which will change the color of their card in the collection. 
Additionally, added functions to upload and download musicxml files to and from the database.

File: app/src/main/java/com/github/scribeWizTeam/scribewiz/Activities/BadgeDisplayActivity.kt [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        setContent{
+            badgeDisplayLayout()
+        }
+    }
+
+    @Composable
+    fun badgeDisplayLayout(){
```
Comment (NoeTerrier) [inline app/src/main/java/com/github/scribeWizTeam/scribewiz/Activities/BadgeDisplayActivity.kt:41]:
```suggestion
    fun BadgeDisplayLayout(){
```

File: app/src/main/java/com/github/scribeWizTeam/scribewiz/Activities/BadgeDisplayActivity.kt [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+import com.github.scribeWizTeam.scribewiz.models.UserModel
+
+class BadgeDisplayActivity : ComponentActivity(){
+
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        setContent{
+            badgeDisplayLayout()
```
Comment (NoeTerrier) [inline app/src/main/java/com/github/scribeWizTeam/scribewiz/Activities/BadgeDisplayActivity.kt:36]:
```suggestion
            BadgeDisplayLayout()
```

File: app/src/main/java/com/github/scribeWizTeam/scribewiz/Activities/BadgeDisplayActivity.kt [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+                    },
+                    modifier = Modifier.height(50.dp).width(50.dp).align(Alignment.CenterVertically),
+
+                ) {
+                    Text("<-")
+                }
+                Text("My badges", fontSize = 20.sp, textAlign = TextAlign.Center,
+                modifier = Modifier.padding(start = 100.dp).align(Alignment.CenterVertically))
+            }
+            displayBadges()
```
Comment (NoeTerrier) [inline app/src/main/java/com/github/scribeWizTeam/scribewiz/Activities/BadgeDisplayActivity.kt:65]:
```suggestion
            Badges()
```

File: app/src/main/java/com/github/scribeWizTeam/scribewiz/Activities/BadgeDisplayActivity.kt [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+
+    private val cardColors = mapOf<Int, Color>(
+        Pair(BadgeRanks.GOLD.ordinal, Color(215,183,64)),
+        Pair(BadgeRanks.SILVER.ordinal, Color(224, 224, 224)),
+        Pair(BadgeRanks.BRONZE.ordinal, Color(184, 115, 51))
+        )
+    // Displays the user's current badges in a grid
+    @Composable
+    fun displayBadges(){
```
Comment (NoeTerrier) [inline app/src/main/java/com/github/scribeWizTeam/scribewiz/Activities/BadgeDisplayActivity.kt:78]:
```suggestion
    fun Badges(){
```

File: app/src/main/java/com/github/scribeWizTeam/scribewiz/models/BadgeModel.kt [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+         */
+        fun getAllBadgesFromUser(user: UserModel) : MutableSet<BadgeModel> {
+            val db = Firebase.firestore
+
+            val badges : MutableSet<BadgeModel> = mutableSetOf()
+
+            runBlocking {
+                val job = launch {
+                    for (id in user.badges!!) {
+                        db.collection("Users")
```
Comment (NoeTerrier) [inline app/src/main/java/com/github/scribeWizTeam/scribewiz/models/BadgeModel.kt:42]:
```suggestion
                        db.collection(UserModel.COLLECTION)
```

File: app/src/main/java/com/github/scribeWizTeam/scribewiz/models/BadgeModel.kt [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        /**
+         * Adds the given badge to the provided user.
+         * If no badge is provided, a default test badge will be added
+         * To get the user, LocalContext.current can be used
+         */
+        fun addBadgeToUser(user: UserModel, badge: BadgeModel?){
+            val time = Calendar.getInstance().time
+            val formatter = DateFormat.getDateInstance()
+
+            var badgeData = BadgeModel(
```
Comment (NoeTerrier) [inline app/src/main/java/com/github/scribeWizTeam/scribewiz/models/BadgeModel.kt:67]: you can put it as a default value in the parameters of the function instead, 
Like:
```
fun addBadgeToUser(user: UserModel, badge: BadgeModel = BadgeModel(...)) {
...
```

File: app/src/main/java/com/github/scribeWizTeam/scribewiz/models/BadgeModel.kt [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+                BadgeRanks.GOLD.ordinal
+            )
+            if(badge != null){
+                badgeData = badge
+                badgeData.dateObtained = formatter.format(time)
+            }
+
+            val userDoc : DocumentReference =
+                Firebase.firestore
+                .collection("Users")
```
Comment (NoeTerrier) [inline app/src/main/java/com/github/scribeWizTeam/scribewiz/models/BadgeModel.kt:82]:
```suggestion
                .collection(UserModel.COLLECTION)
```

File: app/src/main/java/com/github/scribeWizTeam/scribewiz/models/BadgeModel.kt [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+                .document(user.id!!)
+
+            // Add badge in badge collection
+            userDoc.collection("Badges")
```
Comment (NoeTerrier) [inline app/src/main/java/com/github/scribeWizTeam/scribewiz/models/BadgeModel.kt:86]:
```suggestion
            userDoc.collection(COLLECTION)
```

**Review by NoeTerrier — CHANGES_REQUESTED** [review]
May need to do minor changes before merging :)

**Review by Netzu00 — COMMENTED** [review]
Thank you for the review! I've made the changes :)

**Review by NoeTerrier — APPROVED** [review]

**Pull request merged**